### ISSUE TITLE
Fix failing tests (1036 tests completed, 284 failed)

### ISSUE THREAD

**Issue by wololock** [GPars/GPars#48]
Affected branch: master
Running tests result: 1036 tests completed, 284 failed
![gpars_tests_status](https://user-images.githubusercontent.com/1258405/27983876-0bc7cd7c-63c8-11e7-806b-d98276c799ea.png)
Output: https://gist.github.com/wololock/a640ec860e4088b0e5e8d44aa7c0a160

I've already found out that many of these tests fail due to missing methods in classes after refactoring and those missing methods are still being invoked. For example `GParsPoolUtilHelper.retrieveFJPool()` is missing. 

@russel what do you think about treating this issue as an epic and creating fine-grained issues that describe the reason of failure in details (like this missing `GParsPoolUtilHelper.retrieveFJPool()` case). It would help discussing and finding the best way to fix specific case. Here we could trace how many failing tests remain to see the progress. Also using a backlog (https://github.com/GPars/GPars/projects) would be a nice tool to have.

**Comment by russel** [comment]
We need to separate tests relying on ParallelArray which need to go or be replace by Stream based equivalents, and tests which rely on reasonable methods that may have accidentally been removed to get a compile.

ParallelArray is a feature of extra166 which we originally forked but must now remove since JDK8 Stream is the way forward for all data parallel operations on JDK8 and later. JDK8 is the base version for GPars 2.

Functions accessing the ForkJoin pool are accessing something that remains in JDK8 and beyond. Thus these are decisions that really need to be addressed on a case-by-case basis.

Some people have suggested keeping some of the GPars methods that relied on ParallelArray replacing them with Streams versions. So rather than remove, commenting out has been used wherever possible for now.

**Comment by wololock** [comment]
Yes, that's a good idea. I just saw `GParsPoolUtil` class has most of parallel methods commented out and a few is implemented using Java Stream API. I can start implementing those methods if you don't mind (I will create a separate issue for that). One question: what is the desired parallel operations implementation? Some methods use Java Stream API only, e.g.:

```
    public static <T> Collection<T> eachParallel(final Collection<T> collection, final Closure<?> cl) {
        collection.parallelStream().forEach(new ClosureConsumer<T>(cl));
        //GParsPoolUtilHelper.eachParallelPA(GParsPoolUtilHelper.createPAFromCollection(collection, retrievePool()), cl);
        return collection;
    }
```
and some of them use ForkJoinPool + Java Stream API e.g.
```
    public static <K, V> Map<K, V> eachParallel(final Map<K, V> collection, final Closure<V> cl) throws InterruptedException, ExecutionException {
        return retrievePool().submit(() ->
                collection.entrySet().parallelStream().collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> valueFromClosure(e, cl))))
                .get();
    }
```
What is the desired approach to follow? I assume that `retrievePool()` method allows to control the pool size while Java's `parallelStream()` is out of control in this area. Is it correct?

**Comment by russel** [comment]
Hummm… I am not sure. On the one hand just using Streams directly is easier, so on that grounds to be preferred. But if there is a definite benefit to the retrievePool approach we need to find out and document why. This way we have a rationale for an idiom we can impose throughout GPars.

**Comment by wololock** [comment]
I've found one significant difference so far. Consider executing following script:
```
        withPool(2) {
            [1,2,3,4,5,6].eachParallel {
                println("[${Thread.currentThread().name}] ${it}")
            }
        }
```
If `eachParallel` method is implemented like this:
```
    public static <T> Collection<T> eachParallel(final Collection<T> collection, final Closure<?> cl) throws ExecutionException, InterruptedException {
        collection.parallelStream().forEach(new ClosureConsumer<T>(cl));
        //GParsPoolUtilHelper.eachParallelPA(GParsPoolUtilHelper.createPAFromCollection(collection, retrievePool()), cl);
        return collection;
    }
```
we can expect following output:
```
[ForkJoinPool.commonPool-worker-2] 1
[ForkJoinPool.commonPool-worker-3] 6
[ForkJoinPool.commonPool-worker-4] 5
[ForkJoinPool.commonPool-worker-1] 2
[ForkJoinPool.commonPool-worker-5] 3
```
5 different threads were used in this parallel operation, so my parameter in `withPool(2)` was not used. Alternatively if this is my `eachParallel` method:
```
    public static <T> Collection<T> eachParallel(final Collection<T> collection, final Closure<?> cl) throws ExecutionException, InterruptedException {
        return (Collection<T>) retrievePool().submit(() ->
            collection.parallelStream().forEach(new ClosureConsumer<T>(cl))
        ).get();
    }
```
we got following output (only 2 threads used for run parallel operation):
```
[ForkJoinPool-1-worker-1] 4
[ForkJoinPool-1-worker-0] 2
[ForkJoinPool-1-worker-0] 3
[ForkJoinPool-1-worker-1] 6
[ForkJoinPool-1-worker-0] 1
[ForkJoinPool-1-worker-1] 5
```
In this case using `retrievePool().submit()` for triggering parallel operations makes more sense.

**Comment by russel** [comment]
Good piece of evidence based work. Thanks for checking that out. It seems then that there is no choice and that my earlier bits of code are in the wrong idiom – i.e. I should have done this check ages ago. :-)

**Comment by wololock** [comment]
No worries, I've also added a test case to cover this corner case - it checks that running e.g. `withPool(2)` triggers 2 threads at maximum to run computation :)

**Comment by wololock** [comment]
Related pull request: #50

**Comment by wololock** [comment]
Current status from Travis CI:

![przechwycenie obrazu ekranu_2017-07-11_08-34-26](https://user-images.githubusercontent.com/1258405/28054074-cedf64d0-6613-11e7-8ec1-67a98e83316e.png)

**Comment by russel** [comment]
Current 2018-01-03 status on my local machine is 25 failures. 4 of these are real, 21 are likely to go away as the tests are testing redundant functionality that is likely to be removed in the next few days.

**Comment by russel** [comment]
Status as at 2018-01-09 using JDK9 it seems there are 5 test fails and all to do with the getParallel function, albeit possibly indirectly.
